monai/networks/nets/masked_autoencoder_vit.py

```diff
@@ -0,0 +1,221 @@
+        mlp_dim: int = 512,
+        num_layers: int = 12,
+        num_heads: int = 12,
+        masking_ratio: float = 0.75,
+        decoder_hidden_size: int = 384,
+        decoder_mlp_dim: int = 512,
+        decoder_num_layers: int = 4,
+        decoder_num_heads: int = 12,
+        proj_type: str = "conv",
+        pos_embed_type: str = "sincos",
+        decoder_pos_embed_type: str = "sincos",
+        dropout_rate: float = 0.0,
+        spatial_dims: int = 3,
+        qkv_bias: bool = False,
+        save_attn: bool = False,
+    ) -> None:
+        """
+        Args:
+            in_channels: dimension of input channels or the number of channels for input.
+            img_size: dimension of input image.
+            patch_size: dimension of patch size
+            hidden_size: dimension of hidden layer. Defaults to 768.
+            mlp_dim: dimension of feedforward layer. Defaults to 512.
+            num_layers:  number of transformer blocks. Defaults to 12.
+            num_heads: number of attention heads. Defaults to 12.
+            masking_ratio: ratio of patches to be masked. Defaults to 0.75.
+            decoder_hidden_size: dimension of hidden layer for decoder. Defaults to 384.
+            decoder_mlp_dim: dimension of feedforward layer for decoder. Defaults to 512.
+            decoder_num_layers: number of transformer blocks for decoder. Defaults to 4.
+            decoder_num_heads: number of attention heads for decoder. Defaults to 12.
+            proj_type: position embedding layer type. Defaults to "conv".
+            pos_embed_type: position embedding layer type. Defaults to "sincos".
+            decoder_pos_embed_type: position embedding layer type for decoder. Defaults to "sincos".
+            dropout_rate: fraction of the input units to drop. Defaults to 0.0.
+            spatial_dims: number of spatial dimensions. Defaults to 3.
+            qkv_bias: apply bias to the qkv linear layer in self attention block. Defaults to False.
+            save_attn: to make accessible the attention in self attention block. Defaults to False.
+        Examples::
+            # for single channel input with image size of (96,96,96), and sin-cos positional encoding
+            >>> net = MaskedAutoEncoderViT(in_channels=1, img_size=(96,96,96), patch_size=(16,16,16),
+            pos_embed_type='sincos')
+            # for 3-channel with image size of (128,128,128) and a learnable positional encoding
+            >>> net = MaskedAutoEncoderViT(in_channels=3, img_size=128, patch_size=16, pos_embed_type='learnable')
+            # for 3-channel with image size of (224,224) and a masking ratio of 0.25
+            >>> net = MaskedAutoEncoderViT(in_channels=3, img_size=(224,224), patch_size=(16,16), masking_ratio=0.25,
+            spatial_dims=2)
+        """
+
+        super().__init__()
+
+        if not (0 <= dropout_rate <= 1):
+            raise ValueError("dropout_rate should be between 0 and 1.")
```

Comment:
Could you please also mention the value in the error message which will show more informations for the users? The same as below. Thanks!
Something like:
```suggestion
            raise ValueError(f"dropout_rate should be between 0 and 1, got {dropout_rate}.")
```

Reply:
@KumoLiu 
Concerning the request for comparison: I've carried out an experiment on CIFAR-10 in 2D with 1000 pre-training epochs.

Here are the validation losses obtained and an example reconstruction: (left) this implementation, (right) official implementation.
<img width="1343" alt="Capture d’écran 2024-10-23 à 10 58 37" src="https://github.com/user-attachments/assets/3a2e7537-2527-461e-9541-6431984d95f0">
<img width="1048" alt="Capture d’écran 2024-10-23 à 10 59 06" src="https://github.com/user-attachments/assets/bfc8b118-40f1-482c-8acb-6b218ab4ebbf">


For a comparison in terms of fine-tuning accuracy and computational performance, see the table below.

|                                | This implem. | Official implem. |
|:------------------------------:|:------------:|:----------------:|
|   N parameters (in millions)   |     15.60    |       15.63      |
|      GPU consumption (MiB)     |     8733     |       9185       |
|      Training Time (hours)     |       9      |         8        |
| 5 epoch Fine-tuning  accuracy |     0.86     |       0.86       |

The same training from scratch gives 0.6 accuracy.

Note that what changes a little is the calculation time, which is longer with this implementation. Indeed, in the official implementation, token selection via shuffling + removing the last part is extremely fast, whereas here boolean indexing is a little slower.
Knowing that this selection is really the part apart from the official code, I haven't kept it for licensing reasons, but a similar process coded differently could be envisaged.
Although, I don't know whether this would be acceptable in terms of license compatibility.